src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -0,0 +1,190 @@
+};
+
+class XdgDecorationManagerV1 : public wayland::XdgDecorationManagerV1
+{
+public:
+    XdgDecorationManagerV1(wl_resource* resource);
+
+    class Global : public wayland::XdgDecorationManagerV1::Global
+    {
+    public:
+        Global(wl_display* display);
+
+    private:
+        void bind(wl_resource* new_zxdg_decoration_manager_v1) override;
+    };
+
+private:
+    void get_toplevel_decoration(wl_resource* id, wl_resource* toplevel) override;
+    std::shared_ptr<ToplevelsWithDecorations> const toplevels_with_decorations;
+};
+
+class XdgToplevelDecorationV1 : public wayland::XdgToplevelDecorationV1
+{
+public:
+    XdgToplevelDecorationV1(wl_resource* id, mir::frontend::XdgToplevelStable* toplevel);
+
+    void set_mode(uint32_t mode) override;
+    void unset_mode() override;
+
+private:
+    void update_mode(uint32_t new_mode);
+
+    static uint32_t const default_mode = Mode::client_side;
+
+    mir::frontend::XdgToplevelStable* toplevel;
+    uint32_t mode;
+};
+} // namespace frontend
+} // namespace mir
+
+auto mir::frontend::create_xdg_decoration_unstable_v1(wl_display* display)
+    -> std::shared_ptr<mir::wayland::XdgDecorationManagerV1::Global>
+{
+    return std::make_shared<XdgDecorationManagerV1::Global>(display);
+}
+
+mir::frontend::XdgDecorationManagerV1::Global::Global(wl_display* display) :
+    wayland::XdgDecorationManagerV1::Global::Global{display, Version<1>{}}
+{
+}
+
+void mir::frontend::XdgDecorationManagerV1::Global::bind(wl_resource* new_zxdg_decoration_manager_v1)
+{
+    new XdgDecorationManagerV1{new_zxdg_decoration_manager_v1};
+}
+
+mir::frontend::XdgDecorationManagerV1::XdgDecorationManagerV1(wl_resource* resource) :
+    mir::wayland::XdgDecorationManagerV1{resource, Version<1>{}},
+    toplevels_with_decorations{std::make_shared<ToplevelsWithDecorations>()}
+{
+}
+
+void mir::frontend::XdgDecorationManagerV1::get_toplevel_decoration(wl_resource* id, wl_resource* toplevel)
+{
+    using Error = mir::frontend::XdgToplevelDecorationV1::Error;
+
+    auto* tl = mir::frontend::XdgToplevelStable::from(toplevel);
+    if (!tl)
+    {
+        BOOST_THROW_EXCEPTION(std::runtime_error("Invalid toplevel pointer"));
+    }
+
+    auto decoration = new XdgToplevelDecorationV1{id, tl};
+    if (!toplevels_with_decorations->register_toplevel(toplevel))
+    {
+        BOOST_THROW_EXCEPTION(mir::wayland::ProtocolError(
+            resource, Error::already_constructed, "Decoration already constructed for this toplevel"));
+    }
+
+    decoration->add_destroy_listener(
+        [toplevels_with_decorations = this->toplevels_with_decorations, toplevel]()
+        {
+            toplevels_with_decorations->unregister_toplevel(toplevel);
+        });
+
+    tl->add_destroy_listener(
+        [toplevels_with_decorations = this->toplevels_with_decorations, client = this->client, toplevel]()
+        {
+            if (!client->is_being_destroyed() && !toplevels_with_decorations->unregister_toplevel(toplevel))
```

Comment:
We probably want to unregister regardless of `is_being_destroyed()`? (`&&` short-circuits)